### ISSUE TITLE
image-size, image-width, image-height doesn't work correctly

### ISSUE THREAD

**Issue by mdreizin** [webpack/less-loader#43]
For instance, we have "sub/variables.less" which contains:

```
@img-width: image-width("../img/test.png");
```

And if try to import "~sub/variables.less" in other file (the `resolve.root` is set correctly) via:

```
@import (reference) "~sub/variables.less";
```

the "less-loader" will throws exception:

```
Module build failed: error evaluating function `image-width`: Cannot 'resolveSync' because the fileSystem is not sync. Use 'resolve'!
 @ <full-path-to-file>/sub/less/variables.less
```

**Comment by jhnns** [comment]
Mhmmm that might be a limitation by webpack @sokra? Why is `resolveSync` not allowed?

**Comment by sokra** [comment]
webpack uses a async filesystem. You must use the async version of resolve.

**Comment by malykhinvi** [comment]
I have the same problem. Is there some workaround for this problem?

**Comment by jhnns** [comment]
@sokra Shouldn't sync resolving be opt-in?

**Comment by sokra** [comment]
`resolveSync` can only be used with a sync filesystem. This never occurs with webpack. It only occurs with enhanced-require when you use a sync require. You must not use `resolveSync` in webpack.

The problem is that [some](https://github.com/less/less.js/blob/2b42605337bcba58dec0299a5d000946b2fc714e/lib/less-node/image-size.js#L28) [parts](https://github.com/less/less.js/blob/d206167c840fe4b20ab22164989714ee07778798/lib/less/functions/data-uri.js#L56) of less.js don't use `loadFile` but only `loadFileSync`. Because you can not allowed to use sync io in a loader this breaks...

Either we use a uncool workaround with multiple less passes, or we fixed the occrences of `loadFileSync` in less.js (might require design changes).

**Comment by next0** [comment]
@sokra @jhnns, what about just use less.js `FileManager` insted of `WebpackFileManager` in this case? You can check this https://github.com/webpack/less-loader/pull/58

**Comment by avivr** [comment]
i'm having the same problem, trying to use *image-height("../some/path/to/image.png")* and get:

```javascript
ERROR in ../~/css-loader!../~/less-loader!./less/main.less
Module build failed: error evaluating function `image-height`: Cannot 'resolveSync' because the fileSystem is not sync. Use 'resolve'!
```

can you please explain how to solve/workaround this, or what i'm doing wrong?
thanks!

**Comment by next0** [comment]
@avivr, as @sokra explain before, the *less.js* use only `loadFileSync` for functions like `data-uri`, `image-size`, `image-width`, `image-height` and etc. *less-loader* use custom `WebpackFileManager` as a plugin for *less.js*. The main issue, what webpack can't resolve sync file loads when running in async mode, but `WebpackFileManager` tells [otherwise here](https://github.com/webpack/less-loader/blob/master/index.js#L97). In case if `supportsSync` will be return false here in async mode, all sync file loads will be managed by native *less.js* and that error will be disappeared.
You can check my PR https://github.com/webpack/less-loader/pull/58 , it should fix your issue.

**Comment by jhnns** [comment]
Fix shipped with `2.2.2`
